internal/controller/state/graph/route_common.go

```diff
@@ -285,34 +285,57 @@ func buildSectionNameRefs(
 	}
 	uniqueSectionsPerGateway := make(map[key]struct{})
 
-	for i, p := range parentRefs {
+	parentRefIndex := 0
+
+	for _, p := range parentRefs {
 		gw := findGatewayForParentRef(p, routeNamespace, gws)
 		if gw == nil {
 			continue
 		}
 
-		var sectionName string
-		if p.SectionName != nil {
-			sectionName = string(*p.SectionName)
-		}
-
 		gwNsName := client.ObjectKeyFromObject(gw.Source)
 		k := key{
-			gwNsName:    gwNsName,
-			sectionName: sectionName,
+			gwNsName: gwNsName,
+		}
+
+		// If there is no section name, we create ParentRefs for each listener in the gateway
+		if p.SectionName == nil {
+			for _, l := range gw.Listeners {
+				sectionName := string(l.Source.Name)
+				k.sectionName = sectionName
+
+				if _, exist := uniqueSectionsPerGateway[k]; exist {
```

Comment:
Implementation looks good, only nit I have is that there is duplicated code here, could pull out into a closure function inside this function.